File: kcd2_dice/game.py
"""Match/turn state machine for the KCD2 dice minigame."""

from __future__ import annotations

from dataclasses import dataclass, field, replace
from enum import Enum
from random import Random
from typing import Optional, Sequence

from kcd2_dice.scoring import has_any_score, score_selection


class Player(str, Enum):
    HUMAN = "human"
    AI = "ai"


class Phase(str, Enum):
    NEED_ROLL = "need_roll"
    SELECTING = "selecting"
    GAME_OVER = "game_over"


@dataclass(frozen=True)
class MatchState:
    target: int = 3000
    scores: dict[Player, int] = field(
        default_factory=lambda: {Player.HUMAN: 0, Player.AI: 0}
    )
    current: Player = Player.HUMAN
    turn_score: int = 0
    dice: tuple[int, ...] = (1, 1, 1, 1, 1, 1)
    held: tuple[bool, ...] = (False, False, False, False, False, False)
    selected: tuple[bool, ...] = (False, False, False, False, False, False)
    phase: Phase = Phase.NEED_ROLL
    winner: Optional[Player] = None
    message: str = "点击掷骰子开始回合"

    @property
    def active_indices(self) -> tuple[int, ...]:
        return tuple(i for i, locked in enumerate(self.held) if not locked)

    @property
    def selected_indices(self) -> tuple[int, ...]:
        return tuple(i for i, on in enumerate(self.selected) if on)

    @property
    def selected_values(self) -> tuple[int, ...]:
        return tuple(self.dice[i] for i in self.selected_indices)

    @property
    def selection_score(self) -> Optional[int]:
        if not self.selected_indices:
            return None
        return score_selection(self.selected_values)

    @property
    def selection_is_valid(self) -> bool:
        return self.selection_score is not None


def new_match(target: int = 3000) -> MatchState:
    return MatchState(target=target)


def _clear_selection(state: MatchState) -> MatchState:
    return replace(state, selected=(False,) * 6)


def _set_selected(state: MatchState, indices: Sequence[int]) -> MatchState:
    selected = [False] * 6
    for idx in indices:
        selected[idx] = True
    return replace(state, selected=tuple(selected))


def toggle_die(state: MatchState, index: int) -> MatchState:
    if state.phase != Phase.SELECTING:
        return state
    if index < 0 or index >= 6 or state.held[index]:
        return state
    selected = list(state.selected)
    selected[index] = not selected[index]
    return replace(state, selected=tuple(selected))


def roll(state: MatchState, rng: Random) -> MatchState:
    """Roll all non-held dice. Busts end the turn with turn_score cleared."""
    if state.phase != Phase.NEED_ROLL:
        return state

    dice = list(state.dice)
    for i in state.active_indices:
        dice[i] = rng.randint(1, 6)
    rolled = replace(
        state,
        dice=tuple(dice),
        selected=(False,) * 6,
        phase=Phase.SELECTING,
    )

    active_values = [rolled.dice[i] for i in rolled.active_indices]
    if not has_any_score(active_values):
        return _end_turn_after_bust(rolled)

    return replace(rolled, message="请选择得分骰子")


def _end_turn_after_bust(state: MatchState) -> MatchState:
    nxt = Player.AI if state.current == Player.HUMAN else Player.HUMAN
    return replace(
        state,
        turn_score=0,
        held=(False,) * 6,
        selected=(False,) * 6,
        current=nxt,
        phase=Phase.NEED_ROLL,
        message=f"{'你' if state.current == Player.HUMAN else '对手'}爆点，本轮得分清零",
    )


def _apply_selection(state: MatchState) -> MatchState:
    score = state.selection_score
    if score is None:
        return state

    held = list(state.held)
    for i in state.selected_indices:
        held[i] = True

    turn_score = state.turn_score + score
    # Hot dice: every die contributed this throw → unlock all six.
    if all(held):
        held = [False] * 6

    return replace(
        state,
        turn_score=turn_score,
        held=tuple(held),
        selected=(False,) * 6,
        message=f"本轮累计 {turn_score} 分",
    )


def score_and_continue(state: MatchState, rng: Random) -> MatchState:
    if state.phase != Phase.SELECTING or not state.selection_is_valid:
        return state
    applied = _apply_selection(state)
    return roll(replace(applied, phase=Phase.NEED_ROLL), rng)


def score_and_pass(state: MatchState) -> MatchState:
    if state.phase != Phase.SELECTING or not state.selection_is_valid:
        return state

    applied = _apply_selection(state)
    scores = dict(applied.scores)
    scores[applied.current] = scores[applied.current] + applied.turn_score

    if scores[applied.current] >= applied.target:
        return replace(
            applied,
            scores=scores,
            turn_score=0,
            held=(False,) * 6,
            selected=(False,) * 6,
            phase=Phase.GAME_OVER,
            winner=applied.current,
            message=f"{'你' if applied.current == Player.HUMAN else '对手'}获胜！",
        )

    nxt = Player.AI if applied.current == Player.HUMAN else Player.HUMAN
    return replace(
        applied,
        scores=scores,
        turn_score=0,
        held=(False,) * 6,
        selected=(False,) * 6,
        current=nxt,
        phase=Phase.NEED_ROLL,
        message="回合结束，交换玩家",
    )


def set_selection(state: MatchState, indices: Sequence[int]) -> MatchState:
    """Replace the current selection (used by AI)."""
    if state.phase != Phase.SELECTING:
        return state
    cleaned = []
    for idx in indices:
        if 0 <= idx < 6 and not state.held[idx]:
            cleaned.append(idx)
    return _set_selected(state, cleaned)

File: web/js/game.js
/** Match / turn state machine for KCD2 dice. */

import { hasAnyScore, scoreSelection } from "./scoring.js";

export const Player = { HUMAN: "human", AI: "ai" };
export const Phase = {
  NEED_ROLL: "need_roll",
  SELECTING: "selecting",
  GAME_OVER: "game_over",
};

export function newMatch(target = 3000) {
  return {
    target,
    scores: { [Player.HUMAN]: 0, [Player.AI]: 0 },
    current: Player.HUMAN,
    turnScore: 0,
    dice: [1, 1, 1, 1, 1, 1],
    held: [false, false, false, false, false, false],
    selected: [false, false, false, false, false, false],
    phase: Phase.NEED_ROLL,
    winner: null,
    message: "点「掷骰子」开始",
  };
}

function clone(state) {
  return {
    ...state,
    scores: { ...state.scores },
    dice: [...state.dice],
    held: [...state.held],
    selected: [...state.selected],
  };
}

export function activeIndices(state) {
  return state.held.map((h, i) => (!h ? i : -1)).filter((i) => i >= 0);
}

export function selectedIndices(state) {
  return state.selected.map((on, i) => (on ? i : -1)).filter((i) => i >= 0);
}

export function selectionScore(state) {
  const idx = selectedIndices(state);
  if (!idx.length) return null;
  return scoreSelection(idx.map((i) => state.dice[i]));
}

export function selectionValid(state) {
  return selectionScore(state) !== null;
}

export function toggleDie(state, index) {
  const s = clone(state);
  if (s.phase !== Phase.SELECTING || s.held[index]) return s;
  s.selected[index] = !s.selected[index];
  return s;
}

export function setSelection(state, indices) {
  const s = clone(state);
  if (s.phase !== Phase.SELECTING) return s;
  s.selected = s.selected.map(() => false);
  for (const i of indices) {
    if (!s.held[i]) s.selected[i] = true;
  }
  return s;
}

function endTurnAfterBust(state) {
  const s = clone(state);
  const loser = s.current;
  s.turnScore = 0;
  s.held = s.held.map(() => false);
  s.selected = s.selected.map(() => false);
  s.current = loser === Player.HUMAN ? Player.AI : Player.HUMAN;
  s.phase = Phase.NEED_ROLL;
  s.message = loser === Player.HUMAN ? "你爆点了，本轮清零" : "对手爆点，本轮清零";
  return s;
}

export function roll(state, rng = Math.random) {
  let s = clone(state);
  if (s.phase !== Phase.NEED_ROLL) return s;

  const active = activeIndices(s);
  for (const i of active) {
    s.dice[i] = 1 + Math.floor(rng() * 6);
  }
  s.selected = s.selected.map(() => false);
  s.phase = Phase.SELECTING;

  const values = active.map((i) => s.dice[i]);
  if (!hasAnyScore(values)) return endTurnAfterBust(s);

  s.message = "选出得分骰子";
  return s;
}

function applySelection(state) {
  const score = selectionScore(state);
  if (score === null) return state;
  const s = clone(state);
  for (let i = 0; i < 6; i++) {
    if (s.selected[i]) s.held[i] = true;
  }
  s.turnScore += score;
  if (s.held.every(Boolean)) s.held = s.held.map(() => false); // hot dice
  s.selected = s.selected.map(() => false);
  s.message = `本轮累计 ${s.turnScore}`;
  return s;
}

export function scoreAndContinue(state, rng = Math.random) {
  if (state.phase !== Phase.SELECTING || !selectionValid(state)) return state;
  let s = applySelection(state);
  s.phase = Phase.NEED_ROLL;
  return roll(s, rng);
}

export function scoreAndPass(state) {
  if (state.phase !== Phase.SELECTING || !selectionValid(state)) return state;
  let s = applySelection(state);
  s.scores[s.current] += s.turnScore;

  if (s.scores[s.current] >= s.target) {
    s.winner = s.current;
    s.phase = Phase.GAME_OVER;
    s.turnScore = 0;
    s.held = s.held.map(() => false);
    s.selected = s.selected.map(() => false);
    s.message = s.winner === Player.HUMAN ? "你赢了！" : "对手获胜";
    return s;
  }

  const prev = s.current;
  s.turnScore = 0;
  s.held = s.held.map(() => false);
  s.selected = s.selected.map(() => false);
  s.current = prev === Player.HUMAN ? Player.AI : Player.HUMAN;
  s.phase = Phase.NEED_ROLL;
  s.message = "回合结束";
  return s;
}
